crates/bevy_reflect/src/lib.rs

```diff
@@ -1927,6 +1927,73 @@ bevy_reflect::tests::Test {
         );
     }
 
+    #[test]
+    fn dynamic_types_debug_format() {
+        #[derive(Debug, Reflect)]
+        struct TestTupleStruct(u32);
+
+        #[derive(Debug, Reflect)]
+        enum TestEnum {
+            A(u32),
+            B,
+        }
+
+        #[derive(Debug, Reflect)]
+        // test DynamicStruct
+        struct TestStruct {
+            // test DynamicTuple
+            tuple: (u32, u32),
+            // test DynamicTupleStruct
+            tuple_struct: TestTupleStruct,
+            // test DynamicList
+            list: Vec<u32>,
+            // test DynamicArray
+            array: [u32; 3],
+            // test DynamicEnum
+            e: TestEnum,
+            // test DynamicMap
+            map: HashMap<u32, u32>,
+            // test reflected value
+            value: u32,
+        }
+        let mut map = HashMap::new();
+        map.insert(9, 10);
+        let mut test_struct = TestStruct {
+            tuple: (0, 1),
+            list: vec![2, 3, 4],
+            array: [5, 6, 7],
+            tuple_struct: TestTupleStruct(8),
+            e: TestEnum::A(11),
+            map,
+            value: 12,
+        }
+        .clone_value();
+        let test_struct = test_struct.downcast_mut::<DynamicStruct>().unwrap();
+
+        // test unknown DynamicStruct
+        let mut test_unknown_struct = DynamicStruct::default();
+        test_unknown_struct.insert("a", 13);
+        test_struct.insert("unknown_struct", test_unknown_struct);
+        // test unknown DynamicTupleStruct
+        let mut test_unknown_tuple_struct = DynamicTupleStruct::default();
+        test_unknown_tuple_struct.insert(14);
+        test_struct.insert("unknown_tuplestruct", test_unknown_tuple_struct);
+        assert_eq!(
+            format!("{:?}", test_struct),
+            "DynamicStruct(bevy_reflect::tests::TestStruct { \
+                tuple: DynamicTuple((0, 1)), \
+                tuple_struct: DynamicTupleStruct(bevy_reflect::tests::TestTupleStruct(8)), \
+                list: DynamicList([2, 3, 4]), \
+                array: DynamicArray([5, 6, 7]), \
+                e: DynamicEnum(A(11)), \
+                map: DynamicMap({9: 10}), \
+                value: 12, \
+                unknown_struct: DynamicStruct(Unknown { a: 13 }), \
```

Comment:
> `Unknown` may get confused with some kind of user defined struct that's literally named "Unknown".

Yeah technically a user-defined `Unknown` would have a path prefix. However, I imagine that could be easily overlooked which is why I think it could be worth changing.

> * or `???` if missing type registration is a bug in user code which people have to notice and fix

This wouldn't be a bug. This is just what happens when you create a dynamic type without specifying the type it proxies (`clone_value`/`clone_dynamic` will always pass along that type info).

My preference is for the empty string. I feel like it better indicates that the dynamic is "type-less" and is a little less visually jarring. `???` could work as well, but only if we really wanted to draw attention to the missing proxy.